File: lab2/java/concurrent/Sum.java
import java.io.FileInputStream;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.concurrent.Semaphore;
import java.util.ArrayList;
import java.util.List;

public class Sum {
    private static int globalSum = 0;
    private static final Semaphore semaphore = new Semaphore(1);

    public static void main(String[] args) throws Exception {
        if (args.length < 1) {
            System.err.println("Usage: java Sum filepath1 filepath2 filepathN");
            System.exit(1);
        }

	List<Thread> threads = new ArrayList<>();

	    //many exceptions could be thrown here. we don't care
        for (String path : args) {
            Thread thread = new Thread(new FileSum(path), path);
	    threads.add(thread);
            thread.start();
        }

	for (Thread thread : threads) {
		thread.join();
	}

	System.out.println("Soma total:" + Integer.toString(globalSum));
    }

    public static class FileSum implements Runnable {

        private final String path;
        
        public FileSum(String path) {
            this.path = path;
        }

        @Override
        public void run() {
            try {
                sum();
            } catch (IOException ex) {
            }
        }

        public void sum() throws IOException {
            Path filePath = Paths.get(path);
            if (Files.isRegularFile(filePath)) {
                FileInputStream fis = new FileInputStream(filePath.toString());
		try {
		semaphore.acquire();
		globalSum += sum(fis);
		semaphore.release();
		} catch (Exception ex){}

            } else {
                throw new RuntimeException("Non-regular file: " + path);
            }
        }

        public int sum(FileInputStream fis) throws IOException {
            int byteRead;
            int sum = 0;

            while ((byteRead = fis.read()) != -1) {
                sum += byteRead;
            }


            return sum;
        }

    }
}

File: lab2/python/concurrent/sum.py
import sys
import threading

sum = 0
semaphore = threading.Semaphore(1)

def do_sum(path):
    _sum = 0
    global sum
    with open(path, 'rb') as f:
        byte = f.read(1)
        while byte:
            _sum += int.from_bytes(byte, byteorder='big', signed=False)
            byte = f.read(1)
    # controle de admissão
    semaphore.acquire()
    sum += _sum
    semaphore.release()

def controlled_sum(path):
   semaphore_2.acquire()
   do_sum(path)
   semaphore_2.release()


#many error could be raised error. we don't care       
if __name__ == "__main__":
    paths = sys.argv[1:]
    threads = []

    semaphore_2 = threading.Semaphore(max(1, len(paths)) // 2)

    for path in paths:
        try:
            thread = threading.Thread(target=controlled_sum, args=(path,))
            threads.append(thread)
            thread.start()
        except Exception as e:
            print(f"Erro ao processar {path}: {e}")
    
    for thread in threads:
          thread.join()
    print(sum)
